_posts/2020-06-18-Metal3-dev-env-BareMetal-Cluster-Deployment.md

```diff
@@ -0,0 +1,396 @@
+| Metal3MachineTemplate | Metal3 resource which acts as a template when creating a Controlplane or a Worker node |
+| KubeadmConfigTemplate | A template of `KubeadmConfig`, for Workers, used to generate KubeadmConfig when a new worker node is provisioned |
+
+> **Note** : The corresponding `KubeadmConfig` is copied to the Controlplane/Worker at the time of provisioning.  
+  
+<br/>
+
+
+## BareMetal Cluster Deployment
+
+The deployment scripts primarily use ansible and the existing Kubernetes management cluster (based on minikube ) for deploying the bare-metal cluster. Make sure that some of the environment variables used for Metal³ deployment are set, if you didn't use `config_${user}.sh` for setting the environment variables.
+
+| Parameter           | Description                  | Default                  |
+| ------------------- | ---------------------------- | ------------------------ |
+| CAPI_VERSION        | Version of Metal3 API4       | v1alpha3        |
+| POD_CIDR            | Pod Network CIDR             | 192.168.0.0/18           |
+| CLUSTER_NAME        | Name of bare metal cluster   | test1                    |
+
+<br/>
+
+
+### Steps Involved
+
+All the scripts for cluster provisioning or deprovisioning are located at - [`${metal3-dev-env}/scripts/v1alphaX/`](https://github.com/metal3-io/metal3-dev-env/tree/master/scripts/v1alphaX). The scripts call a common playbook which handles all the tasks that are available.
+
+> info "Note"
+> Here is a depiction of the common steps, mainly involving generating templates and config files. 
+
+<br/>
+
+!["A diagram depicting the Generate Templates workflow"](/assets/2020-06-18-Metal3-dev-env-BareMetal-Cluster-Deployment/metal3-generate-templates.png)
+<br/>
+<br/>
+
+The steps involved in the process are :
+
+- The script calls an ansible playbook with necessary parameter ( from env variables and defaults )
+- The playbook executes the role -, [`${metal3-dev-env}/vm-setup/roles/v1aX_integration_test`](https://github.com/metal3-io/metal3-dev-env/tree/master/vm-setup/roles/v1aX_integration_test), which runs the corresponding [task_file](https://github.com/metal3-io/metal3-dev-env/tree/master/vm-setup/roles/v1aX_integration_test/tasks) for provisioning/deprovisioning the cluster/controlplane or a worker
+- There are [templates](https://github.com/metal3-io/metal3-dev-env/tree/master/vm-setup/roles/v1aX_integration_test/templates) in the role, which are copied to the `manifest` directory. They are used for generating configuration for the cluster and kubeadm, which is then supplied to the Kubernetes module of ansible to create the cluster.
+- During provisioning, first the `clusterctl` env file is generated, then and `manifest` file is created/updated using the `clusterctl` tool, which is then supplied to `kustomize` tool to create the config files for provisioning the resource, from the available templates.
```

Comment:
this is a bit unclear. I modified the playbook so that it now renders the environment variables file and the cluster, then uses clusterctl to generate the full manifest. The rest of the playbook will just break down the manifest into smaller chunks, to deploy a cluster, the controlplane or the workers. However, I am planning to rework this, because it is done in a way that is a bit clumsy, first generating the whole manifest and then using kustomize to break it down into pieces. My goal is to directly generate the bits that we want, i.e. break the template into smaller pieces to not have to use Kustomize there. Maybe I can first push that change, and then we can come back to this here ?

Reply:
@maelk Yes, its indeed a bit unclear. And the changes you are proposing look much clear. I just have a couple of queries - will that require to rework some of the previous sections as well, and how much time will it take.

Reply:
it should not, and I can work on that tomorrow, hopefully it will be merged by end of day tomorrow